<a href="https://colab.research.google.com/github/pujaroy280/DATA608Story3/blob/main/DATA608Story3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Objective

The CDC publishes firearm mortality for each State per 100,000 persons https://www.cdc.gov/nchs/pressroom/sosmap/firearm_mortality/firearm.htm. Each State' firearm control laws can be categorized as very strict to very lax. The purpose of this Story is to answer the question, " Do stricter firearm control laws help reduce firearm mortality?"

The purpose of this assignment is to:

1.   Access the firearm mortality data from the CDC using an available API (https://open.cdc.gov/apis.html)
2.   Create a 5 point Likert scale categorizing gun control laws from most lax to strictest and assign each state to the most appropriate Likert bin.
3.   Determine wether stricter gun control laws result in reduced gun violence deaths.
4.  Present your story using  heat maps.

## Import Python libraries & Retrieve the Data from the API

For this assignment, I retrieved the data from the Centers for Disease Control and Prevention (CDC) API from this link: https://data.cdc.gov/

Then, I browsed and clicked on the Injury & Violence category and searched for firearm mortality data. 1 result appeared which was listed as "NCHS - VSRR Quarterly provisional estimates for selected indicators of mortality" from this link: https://data.cdc.gov/browse?q=firearm%20mortality%20data&sortBy=relevance. I obtained the API endpoint from this link: https://dev.socrata.com/foundry/data.cdc.gov/489q-934x . I finally fetched the data using response requests available from the Python libraries in JSON format instead of CSV.

In [14]:
import requests  # Importing the requests library to make HTTP requests
import pandas as pd  # Importing pandas library for handling data as DataFrame

url = "https://data.cdc.gov/resource/489q-934x.json"  # URL to fetch data from

response = requests.get(url)  # Sending a GET request to the URL and storing the response

if response.status_code == 200:  # Checking if the response status code is 200 (indicating success)
    data = response.json()  # Converting the JSON response to Python data (a list or dictionary)
    df_mortality = pd.DataFrame(data)  # Creating a DataFrame from the JSON data
else:
    print("Error: Unable to retrieve data from the URL.")  # Printing an error message if request fails

print(df_mortality.head())  # Printing the first few rows of the DataFrame

  year_and_quarter                    time_period  \
0          2021 Q1  12 months ending with quarter   
1          2021 Q1  12 months ending with quarter   
2          2021 Q1  12 months ending with quarter   
3          2021 Q1  12 months ending with quarter   
4          2021 Q1  12 months ending with quarter   

                        cause_of_death     rate_type                unit  \
0                           All causes  Age-adjusted  Deaths per 100,000   
1                    Alzheimer disease  Age-adjusted  Deaths per 100,000   
2                             COVID-19  Age-adjusted  Deaths per 100,000   
3                               Cancer  Age-adjusted  Deaths per 100,000   
4  Chronic liver disease and cirrhosis  Age-adjusted  Deaths per 100,000   

  rate_overall rate_sex_female rate_sex_male rate_alaska rate_alabama  ...  \
0        866.3           716.3        1040.4       779.2       1123.4  ...   
1         32.1            36.8          24.8        28.2         51.

In [15]:
df_mortality

year_and_quarter                    time_period  \
0            2021 Q1  12 months ending with quarter   
1            2021 Q1  12 months ending with quarter   
2            2021 Q1  12 months ending with quarter   
3            2021 Q1  12 months ending with quarter   
4            2021 Q1  12 months ending with quarter   
..               ...                            ...   
875          2023 Q2                 3-month period   
876          2023 Q2                 3-month period   
877          2023 Q2                 3-month period   
878          2023 Q2                 3-month period   
879          2023 Q2                 3-month period   

                            cause_of_death     rate_type                unit  \
0                               All causes  Age-adjusted  Deaths per 100,000   
1                        Alzheimer disease  Age-adjusted  Deaths per 100,000   
2                                 COVID-19  Age-adjusted  Deaths per 100,000   
3                                   Cancer  Age-adjusted  Deaths per 100,000   
4      Chronic liver disease and cirrhosis  Age-adjusted  Deaths per 100,000   
..                                     ...           ...                 ...   
875  Pneumonitis due to solids and liquids         Crude  Deaths per 100,000   
876                             Septicemia         Crude  Deaths per 100,000   
877                                 Stroke         Crude  Deaths per 100,000   
878                                Suicide         Crude  Deaths per 100,000   
879                 Unintentional injuries         Crude  Deaths per 100,000   

    rate_overall rate_sex_female rate_sex_male rate_alaska rate_alabama  ...  \
0          866.3           716.3        1040.4       779.2       1123.4  ...   
1           32.1            36.8          24.8        28.2         51.2  ...   
2          120.7              94         153.9        44.4        160.2  ...   
3            142           122.8         167.7         143        160.5  ...   
4           13.9             9.8          18.3        23.6         17.2  ...   
..           ...             ...           ...         ...          ...  ...   
875          5.7             4.8           6.6         2.7          4.5  ...   
876         11.9            11.9            12         9.9         21.4  ...   
877           47            52.7          41.2        19.7         60.2  ...   
878          NaN             NaN           NaN         NaN          NaN  ...   
879          NaN             NaN           NaN         NaN          NaN  ...   

    rate_age_1_4 rate_age_5_14 rate_age_15_24 rate_age_25_34 rate_age_35_44  \
0            NaN           NaN            NaN            NaN            NaN   
1            NaN           NaN            NaN            NaN            NaN   
2            NaN           NaN            NaN            NaN            NaN   
3            NaN           NaN            NaN            NaN            NaN   
4            NaN           NaN            NaN            NaN            NaN   
..           ...           ...            ...            ...            ...   
875          NaN           NaN            0.1            0.3            0.5   
876          0.4           0.2            0.3            0.8            2.2   
877          0.4           0.2            0.4            1.2            4.8   
878          NaN           NaN            NaN            NaN            NaN   
879          NaN           NaN            NaN            NaN            NaN   

    rate_age_45_54 rate_age_55_64 rate_65_74 rate_age_75_84 rate_age_85_plus  
0              NaN            NaN        NaN            NaN              NaN  
1              NaN            NaN        NaN            NaN              NaN  
2              NaN            NaN        NaN            NaN              NaN  
3              NaN            NaN        NaN            NaN              NaN  
4              NaN            NaN        NaN            NaN              NaN 

Once I successfully fetched the data from the CDC API, I filter the dataframe to observe the mortality rates for only "Firearm-related injury" and the type as "Crude". For the time period, I selected 12 months ending with quarter.

In [16]:
# Filtering DataFrame based on certain conditions
df_gun = df_mortality[df_mortality['cause_of_death'] == "Firearm-related injury"]
df_gun = df_gun[df_gun['rate_type'] == "Crude"]
df_gun = df_gun[df_gun['time_period'] == "12 months ending with quarter"]

# Creating DataFrame with row names as NULL
df_gun = pd.DataFrame(df_gun.reset_index(drop=True))
df_gun

year_and_quarter                    time_period          cause_of_death  \
0          2021 Q1  12 months ending with quarter  Firearm-related injury   
1          2021 Q2  12 months ending with quarter  Firearm-related injury   
2          2021 Q3  12 months ending with quarter  Firearm-related injury   
3          2021 Q4  12 months ending with quarter  Firearm-related injury   
4          2022 Q1  12 months ending with quarter  Firearm-related injury   
5          2022 Q2  12 months ending with quarter  Firearm-related injury   
6          2022 Q3  12 months ending with quarter  Firearm-related injury   
7          2022 Q4  12 months ending with quarter  Firearm-related injury   
8          2023 Q1  12 months ending with quarter  Firearm-related injury   
9          2023 Q2  12 months ending with quarter  Firearm-related injury   

  rate_type                unit rate_overall rate_sex_female rate_sex_male  \
0     Crude  Deaths per 100,000         14.1             3.9          24.5   
1     Crude  Deaths per 100,000         14.4               4            25   
2     Crude  Deaths per 100,000         14.6             4.1          25.3   
3     Crude  Deaths per 100,000         14.7             4.2          25.5   
4     Crude  Deaths per 100,000         14.8             4.1          25.6   
5     Crude  Deaths per 100,000         14.8             4.2          25.6   
6     Crude  Deaths per 100,000         14.7             4.2          25.3   
7     Crude  Deaths per 100,000         14.5             4.1            25   
8     Crude  Deaths per 100,000         14.4             4.1          24.8   
9     Crude  Deaths per 100,000          NaN             NaN           NaN   

  rate_alaska rate_alabama  ... rate_age_1_4 rate_age_5_14 rate_age_15_24  \
0        23.1         24.2  ...          0.8           1.7             23   
1        25.1         24.8  ...          0.8           1.7           23.7   
2        24.4         25.4  ...          0.8           1.7           23.7   
3        24.8         26.1  ...          0.9           1.6           23.5   
4        25.8         25.4  ...            1           1.6           23.3   
5        23.7         25.2  ...            1           1.6           22.4   
6        23.4         25.6  ...          0.9           1.6           21.9   
7        22.4         25.2  ...          0.9           1.5           21.1   
8        21.5         26.1  ...          0.8           1.6             21   
9         NaN          NaN  ...          NaN           NaN            NaN   

  rate_age_25_34 rate_age_35_44 rate_age_45_54 rate_age_55_64 rate_65_74  \
0           23.6           17.3           13.6           11.7       10.9   
1           24.7           17.7           13.8           11.7         11   
2             25             18             14           11.7       11.3   
3           24.8           18.1           14.5           12.1       11.7   
4           24.5           18.4           14.7           12.3       11.7   
5           24.2           18.4           14.9           12.9       12.1   
6           23.6           18.3           14.9           13.1       12.1   
7           22.9           18.1           14.8           13.4       11.9   
8           22.5             18           14.7           13.5         12   
9            NaN            NaN            NaN            NaN        NaN   

  rate_age_75_84 rate_age_85_plus  
0           15.1               16  
1           15.7             17.3  
2           16.1             17.8  
3           16.2             18.3  
4           16.3             19.2  
5           16.6             18.1  
6           16.3             18.6  
7           16.5             18.5  
8           16.3             17.7  
9            NaN              NaN  

[10 rows x 69 columns]

I transformed the data by mapping states as abbreviations to make the data look simple and presentable in the heatmap.

In [17]:
# Mapping state abbreviations to full names
state_abbreviations = {
    "AL": "alabama", "AK": "alaska", "AZ": "arizona", "AR": "arkansas", "CA": "california",
    "CO": "colorado", "CT": "connecticut", "DE": "delaware", "FL": "florida", "GA": "georgia",
    "HI": "hawaii", "ID": "idaho", "IL": "illinois", "IN": "indiana", "IA": "iowa",
    "KS": "kansas", "KY": "kentucky", "LA": "louisiana", "ME": "maine", "MD": "maryland",
    "MA": "massachusetts", "MI": "michigan", "MN": "minnesota", "MS": "mississippi", "MO": "missouri",
    "MT": "montana", "NE": "nebraska", "NV": "nevada", "NH": "new_hampshire", "NJ": "new_jersey",
    "NM": "new_mexico", "NY": "new_york", "NC": "north_carolina", "ND": "north_dakota",
    "OH": "ohio", "OK": "oklahoma", "OR": "oregon", "PA": "pennsylvania", "RI": "rhode_island",
    "SC": "south_carolina", "SD": "south_dakota", "TN": "tennessee", "TX": "texas", "UT": "utah",
    "VT": "vermont", "VA": "virginia", "WA": "washington", "WV": "west_virginia", "WI": "wisconsin",
    "WY": "wyoming", "DC": "district_of_columbia"
}

# Looping through each state abbreviation
for abbrev, full_name in state_abbreviations.items():
    pattern = "rate_" + full_name
    df_gun.columns = df_gun.columns.str.replace(pattern, abbrev)

state_abbreviations

{'AL': 'alabama',
 'AK': 'alaska',
 'AZ': 'arizona',
 'AR': 'arkansas',
 'CA': 'california',
 'CO': 'colorado',
 'CT': 'connecticut',
 'DE': 'delaware',
 'FL': 'florida',
 'GA': 'georgia',
 'HI': 'hawaii',
 'ID': 'idaho',
 'IL': 'illinois',
 'IN': 'indiana',
 'IA': 'iowa',
 'KS': 'kansas',
 'KY': 'kentucky',
 'LA': 'louisiana',
 'ME': 'maine',
 'MD': 'maryland',
 'MA': 'massachusetts',
 'MI': 'michigan',
 'MN': 'minnesota',
 'MS': 'mississippi',
 'MO': 'missouri',
 'MT': 'montana',
 'NE': 'nebraska',
 'NV': 'nevada',
 'NH': 'new_hampshire',
 'NJ': 'new_jersey',
 'NM': 'new_mexico',
 'NY': 'new_york',
 'NC': 'north_carolina',
 'ND': 'north_dakota',
 'OH': 'ohio',
 'OK': 'oklahoma',
 'OR': 'oregon',
 'PA': 'pennsylvania',
 'RI': 'rhode_island',
 'SC': 'south_carolina',
 'SD': 'south_dakota',
 'TN': 'tennessee',
 'TX': 'texas',
 'UT': 'utah',
 'VT': 'vermont',
 'VA': 'virginia',
 'WA': 'washington',
 'WV': 'west_virginia',
 'WI': 'wisconsin',
 'WY': 'wyoming',
 'DC': 'district_of_columbia

In [18]:
# Data type conversion: columns 6 to 69 are converted to double.
df_gun.iloc[:, 6:70] = df_gun.iloc[:, 6:70].apply(pd.to_numeric)

# Extracting year from year_and_quarter and grouping by year
df_gun['year'] = df_gun['year_and_quarter'].str[:4]
df_gun_grouped = df_gun.groupby('year')

# Filtering data for the year 2022 Q4
df_gun_2022 = df_gun[df_gun['year_and_quarter'] == "2022 Q4"]

Once I completed extracting year and filtering data, I tidied the data and mapped gun law ranks for each state.

In [19]:
# Pivoting long
df_gun_2022_long = df_gun_2022.melt(id_vars=['year', 'year_and_quarter'],
                                     value_vars=['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL',
                                                 'GA', 'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA',
                                                 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE',
                                                 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI',
                                                 'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VT', 'WA', 'WI', 'WV',
                                                 'WY'],
                                     var_name='state',
                                     value_name='rate')

# Selecting specific columns for final DataFrame
final_df = df_gun_2022_long[['year', 'state', 'rate']]

# Add gun law rank to final_df
final_df['gun_laws'] = final_df['state'].map({
    "AK": "1", "AL": "1", "AR": "1", "AZ": "1", "GA": "1", "IA": "1", "ID": "1", "IN": "1", "KS": "1",
    "KY": "1", "LA": "1", "ME": "1", "MO": "1", "MS": "1", "MT": "1", "ND": "1", "NH": "1", "OH": "1",
    "OK": "1", "SC": "1", "SD": "1", "TN": "1", "TX": "1", "UT": "1", "WV": "1", "WY": "1",
    "WI": "2",
    "FL": "3", "MI": "3", "MN": "3", "NC": "3", "NE": "3", "NM": "3", "NV": "3", "VT": "3",
    "CO": "4", "DE": "4", "OR": "4", "PA": "4", "RI": "4", "VA": "4", "WA": "4",
    "CA": "5", "CT": "5", "DC": "5", "HI": "5", "IL": "5", "MA": "5", "MD": "5", "NJ": "5", "NY": "5"
})

# Converting gun_laws and year to numeric type
final_df['gun_laws'] = pd.to_numeric(final_df['gun_laws'])
final_df['year'] = pd.to_numeric(final_df['year'])

# Displaying final DataFrame
print(final_df)

    year state  rate  gun_laws
0   2022    AK  22.4         1
1   2022    AL  25.2         1
2   2022    AR  21.9         1
3   2022    AZ  20.9         1
4   2022    CA   8.9         5
5   2022    CO  17.7         4
6   2022    CT   6.9         5
7   2022    DC  22.9         5
8   2022    DE  12.2         4
9   2022    FL  14.5         3
10  2022    GA  19.8         1
11  2022    HI   4.6         5
12  2022    IA  11.5         1
13  2022    ID  17.4         1
14  2022    IL  14.3         5
15  2022    IN  17.7         1
16  2022    KS  16.8         1
17  2022    KY  18.6         1
18  2022    LA  27.6         1
19  2022    MA   3.8         5
20  2022    MD  13.2         5
21  2022    ME  12.9         1
22  2022    MI  15.0         3
23  2022    MN   9.8         3
24  2022    MO  24.1         1
25  2022    MS  28.8         1
26  2022    MT  24.4         1
27  2022    NC  17.0         3
28  2022    ND  16.0         1
29  2022    NE  12.4         3
30  2022    NH  11.2         1
31  2022

After that, I created a mapping dictionary for a Likert scale to create a heat map that categorizes based on gun control law ranks. Then, I applied Likert scale categorization based on gun control law ranks.

In [20]:
# Creating a mapping dictionary for Likert scale
likert_scale_mapping = {
    1: "Very Lax",
    2: "Lax",
    3: "Moderate",
    4: "Strict",
    5: "Very Strict"
}

# Applying Likert scale categorization based on gun control law ranks
final_df['likert_scale'] = final_df['gun_laws'].map(likert_scale_mapping)

# Displaying final DataFrame with Likert scale
print(final_df)

    year state  rate  gun_laws likert_scale
0   2022    AK  22.4         1     Very Lax
1   2022    AL  25.2         1     Very Lax
2   2022    AR  21.9         1     Very Lax
3   2022    AZ  20.9         1     Very Lax
4   2022    CA   8.9         5  Very Strict
5   2022    CO  17.7         4       Strict
6   2022    CT   6.9         5  Very Strict
7   2022    DC  22.9         5  Very Strict
8   2022    DE  12.2         4       Strict
9   2022    FL  14.5         3     Moderate
10  2022    GA  19.8         1     Very Lax
11  2022    HI   4.6         5  Very Strict
12  2022    IA  11.5         1     Very Lax
13  2022    ID  17.4         1     Very Lax
14  2022    IL  14.3         5  Very Strict
15  2022    IN  17.7         1     Very Lax
16  2022    KS  16.8         1     Very Lax
17  2022    KY  18.6         1     Very Lax
18  2022    LA  27.6         1     Very Lax
19  2022    MA   3.8         5  Very Strict
20  2022    MD  13.2         5  Very Strict
21  2022    ME  12.9         1  

In [21]:
# Sort the DataFrame by the "rate" column in descending order
final_df_sorted = final_df.sort_values(by='rate', ascending=False)

# Print the sorted DataFrame with Likert scale
print(final_df_sorted[['state', 'rate', 'gun_laws']])

   state  rate  gun_laws
25    MS  28.8         1
18    LA  27.6         1
32    NM  27.0         3
1     AL  25.2         1
26    MT  24.4         1
24    MO  24.1         1
7     DC  22.9         5
0     AK  22.4         1
2     AR  21.9         1
50    WY  21.3         1
42    TN  21.0         1
3     AZ  20.9         1
40    SC  20.9         1
10    GA  19.8         1
36    OK  19.8         1
33    NV  19.4         3
17    KY  18.6         1
15    IN  17.7         1
5     CO  17.7         4
49    WV  17.5         1
13    ID  17.4         1
27    NC  17.0         3
16    KS  16.8         1
28    ND  16.0         1
35    OH  15.6         1
41    SD  15.5         1
43    TX  15.4         1
37    OR  15.4         4
45    VA  15.1         4
22    MI  15.0         3
38    PA  15.0         4
9     FL  14.5         3
14    IL  14.3         5
48    WI  14.1         2
44    UT  13.2         1
20    MD  13.2         5
47    WA  13.1         4
46    VT  13.0         3
21    ME  12.9         1


In [22]:
# Sort the DataFrame by the "gun_laws" column in descending order
final_df_sorted = final_df.sort_values(by='gun_laws', ascending=False)

# Print the sorted DataFrame with Likert scale
print(final_df_sorted[['state', 'rate', 'gun_laws']])

   state  rate  gun_laws
14    IL  14.3         5
4     CA   8.9         5
20    MD  13.2         5
6     CT   6.9         5
7     DC  22.9         5
19    MA   3.8         5
34    NY   5.3         5
31    NJ   5.1         5
11    HI   4.6         5
45    VA  15.1         4
38    PA  15.0         4
39    RI   3.4         4
37    OR  15.4         4
47    WA  13.1         4
8     DE  12.2         4
5     CO  17.7         4
27    NC  17.0         3
46    VT  13.0         3
9     FL  14.5         3
29    NE  12.4         3
22    MI  15.0         3
23    MN   9.8         3
32    NM  27.0         3
33    NV  19.4         3
48    WI  14.1         2
35    OH  15.6         1
36    OK  19.8         1
0     AK  22.4         1
40    SC  20.9         1
41    SD  15.5         1
43    TX  15.4         1
44    UT  13.2         1
49    WV  17.5         1
42    TN  21.0         1
25    MS  28.8         1
30    NH  11.2         1
28    ND  16.0         1
26    MT  24.4         1
1     AL  25.2         1


In [23]:
# Sort the DataFrame by the "state" column in descending order
final_df_sorted = final_df.sort_values(by='state', ascending=False)

# Print the sorted DataFrame with Likert scale
print(final_df_sorted[['state', 'rate', 'gun_laws']])

   state  rate  gun_laws
50    WY  21.3         1
49    WV  17.5         1
48    WI  14.1         2
47    WA  13.1         4
46    VT  13.0         3
45    VA  15.1         4
44    UT  13.2         1
43    TX  15.4         1
42    TN  21.0         1
41    SD  15.5         1
40    SC  20.9         1
39    RI   3.4         4
38    PA  15.0         4
37    OR  15.4         4
36    OK  19.8         1
35    OH  15.6         1
34    NY   5.3         5
33    NV  19.4         3
32    NM  27.0         3
31    NJ   5.1         5
30    NH  11.2         1
29    NE  12.4         3
28    ND  16.0         1
27    NC  17.0         3
26    MT  24.4         1
25    MS  28.8         1
24    MO  24.1         1
23    MN   9.8         3
22    MI  15.0         3
21    ME  12.9         1
20    MD  13.2         5
19    MA   3.8         5
18    LA  27.6         1
17    KY  18.6         1
16    KS  16.8         1
15    IN  17.7         1
14    IL  14.3         5
13    ID  17.4         1
12    IA  11.5         1


After that, I created a choropleth interactive heat map to visualize the relationship between gun control laws (rates) and gun violence deaths (mortality rates). I defined the likert scale for mortality rates based on how safe and deadliest gun violence rates are in each state. Likewise, for gun control laws based on rates, I defined the likert scale from most lax to most strict.

In [62]:
import pandas as pd
import plotly.graph_objects as go

# Convert 'rate' column to numeric, replacing any non-numeric values with NaN
final_df['rate'] = pd.to_numeric(final_df['rate'], errors='coerce')

# Plotting choropleth interactive heat map to visualize the relationship between gun control laws and gun violence deaths
fig = go.Figure(data=go.Choropleth(
    locations=final_df['state'],  # States
    z=final_df['rate'],  # Values to be color-coded
    locationmode='USA-states',  # Set plot type to US states
    colorscale='armyrose',  # valid colorscale name
    colorbar=dict(title='Gun Violence Rate (Mortality Rate)'),  # Colorbar title
))

# Update layout of Choropleth Interactive Heat Map: Gun Violence Deaths by State and Gun Control Laws
fig.update_layout(
    title='Gun Violence Deaths by States in USA',
    geo=dict(scope='usa',  # Set map scope to USA
             projection_type='albers usa'),  # Albers USA projection
    xaxis_title='Gun Control Laws (Likert Scale)',
    yaxis_title='State'
)

# Define Likert scale categories with reversed order for gun violence deaths
likert_scale = {
    1: 'Safest',
    2: 'Safe',
    3: 'Moderate',
    4: 'Very Dangerous',
    5: 'Deadliest'
}

# Determine positions for each Likert scale category based on the range of values
likert_positions = {
    category: idx / (len(likert_scale) - 0.7) - 0.01
    for idx, category in enumerate(likert_scale.keys())
}

# Add annotations for Likert scale on the left side
for category, label in likert_scale.items():
    fig.add_annotation(
        x=1.20, y=likert_positions[category],  # Positioning the annotation on the right
        xref='paper', yref='paper',  # Define the reference point
        text=label,  # Text to display
        showarrow=False,  # Don't show arrow
        font=dict(size=12, color='black'),  # Font settings
    )

# Show plot
fig.show()


In [61]:
import pandas as pd
import plotly.graph_objects as go

# Define Likert scale categories based on gun laws ratings of each state
likert_scale = {
    1: 'Most Lax',
    2: 'Lax',
    3: 'Neutral',
    4: 'Strict',
    5: 'Most Strict'
}

# Convert 'rate' column to numeric, replacing any non-numeric values with NaN
final_df['rate'] = pd.to_numeric(final_df['rate'], errors='coerce')

# Plotting choropleth interactive heat map to visualize the relationship between gun control laws and gun violence deaths
fig = go.Figure(data=go.Choropleth(
    locations=final_df['state'],  # States
    z=final_df['gun_laws'],  # Values to be color-coded based on gun laws
    locationmode='USA-states',  # Set plot type to US states
    colorscale='tealrose',  # valid colorscale name
    colorbar=dict(
        title='Gun Law Rank',
        tickvals=list(likert_scale.keys()),
        ticktext=list(likert_scale.values()),
        len=0.75
    ),  # Colorbar configuration
    text=final_df['gun_laws'].map(likert_scale),  # Hover text based on Likert scale
))

# Update layout of Choropleth Interactive Heat Map: Gun Control Laws by State
fig.update_layout(
    title='Gun Control Laws by States in USA',
    geo=dict(scope='usa',  # Set map scope to USA
             projection_type='albers usa'),  # Albers USA projection
    xaxis_title='Gun Control Laws (Likert Scale)',
    yaxis_title='State'
)

# Show plot
fig.show()


## Choropleth Heat Maps based on Gun Violence Deaths (Mortality Rate) & Gun Control Laws (Rank)

In [60]:
import pandas as pd
import plotly.graph_objects as go

# Convert 'rate' column to numeric, replacing any non-numeric values with NaN
final_df['rate'] = pd.to_numeric(final_df['rate'], errors='coerce')

# Plotting choropleth interactive heat map to visualize the relationship between gun control laws and gun violence deaths
fig = go.Figure(data=go.Choropleth(
    locations=final_df['state'],  # States
    z=final_df['rate'],  # Values to be color-coded
    locationmode='USA-states',  # Set plot type to US states
    colorscale='armyrose',  # valid colorscale name
    colorbar=dict(title='Gun Violence Rate (Mortality Rate)'),  # Colorbar title
))

# Update layout of Choropleth Interactive Heat Map: Gun Violence Deaths by State and Gun Control Laws
fig.update_layout(
    title='Gun Violence Deaths by States in USA',
    geo=dict(scope='usa',  # Set map scope to USA
             projection_type='albers usa'),  # Albers USA projection
    xaxis_title='Gun Control Laws (Likert Scale)',
    yaxis_title='State'
)

# Define Likert scale categories with reversed order for gun violence deaths
likert_scale = {
    1: 'Safest',
    2: 'Safe',
    3: 'Moderate',
    4: 'Very Dangerous',
    5: 'Deadliest'
}

# Determine positions for each Likert scale category based on the range of values
likert_positions = {
    category: idx / (len(likert_scale) - 0.7) - 0.01
    for idx, category in enumerate(likert_scale.keys())
}

# Add annotations for Likert scale on the left side
for category, label in likert_scale.items():
    fig.add_annotation(
        x=1.20, y=likert_positions[category],  # Positioning the annotation on the right
        xref='paper', yref='paper',  # Define the reference point
        text=label,  # Text to display
        showarrow=False,  # Don't show arrow
        font=dict(size=12, color='black'),  # Font settings
    )

# Show plot
fig.show()

# SECOND HEATMAP ------------------------------------------------------------------------------------------------------------------------------------------------------------------

# Define Likert scale categories based on gun laws ratings of each state
likert_scale = {
    1: 'Most Lax',
    2: 'Lax',
    3: 'Neutral',
    4: 'Strict',
    5: 'Most Strict'
}

# Convert 'rate' column to numeric, replacing any non-numeric values with NaN
final_df['rate'] = pd.to_numeric(final_df['rate'], errors='coerce')

# Plotting choropleth interactive heat map to visualize the relationship between gun control laws and gun violence deaths
fig = go.Figure(data=go.Choropleth(
    locations=final_df['state'],  # States
    z=final_df['gun_laws'],  # Values to be color-coded based on gun laws
    locationmode='USA-states',  # Set plot type to US states
    colorscale='armyrose',  # valid colorscale name
    colorbar=dict(
        title='Gun Law Rank',
        tickvals=list(likert_scale.keys()),
        ticktext=list(likert_scale.values()),
        len=0.75
    ),  # Colorbar configuration
    text=final_df['gun_laws'].map(likert_scale),  # Hover text based on Likert scale
))

# Update layout of Choropleth Interactive Heat Map: Gun Control Laws by State
fig.update_layout(
    title='Gun Control Laws by States in USA',
    geo=dict(scope='usa',  # Set map scope to USA
             projection_type='albers usa'),  # Albers USA projection
    xaxis_title='Gun Control Laws (Likert Scale)',
    yaxis_title='State'
)

# Show plot
fig.show()



## Conclusion
Based on the data collected and presented in the choropleth heat maps, throughout 2022, residents living in Mississippi were prone; had higher chances of being victims of gun violence deaths since the mortality rate was 28.8 and the gun law rank was most lax; not strict. On the other hand, residents residing in Rhode Island were less likely to become victims of gun violence deaths since the mortality rate was 3.4 and gun law rank was strict. Likewise, Massachusetts had a mortality rate of 3.8 for gun violence deaths and the gun law was the most strict. **Overall, it is clear that states with stricter gun laws have lower rates of gun violence deaths.** This is because the heat maps displays that 8 states including CA, HI, IL, NY, NJ, MD, MA and CT that have most stricter gun law ranks have safest to safe gun violence (mortality) rates based on the Likert scale. On the other hand, 26 states including MT, ND, SD, ID, WY, UT, AZ, AK, IA, MO, KS, OK, TX, AR, LA, IN, OH, WV, KY, TN, MS, AL, GA, SC, NH, ME have a gun law rank of 1 (Most Lax) based on the Likert scale and have moderate to deadliest gun violence (mortality) rates. **Therefore, stricter firearm control laws help reduce firearm mortality.**